### PR TITLE
flow merge for force-triggered n=0 (e.g. queued) tasks

### LINKED ISSUE #6174: triggering a queued task in a different flow has no effect
It would appear that it is not possible to trigger a task in a different flow when it is queued. This could, potentially, be true of waiting states in general, investigation needed.

Example workflow:

```ini
[scheduling]
    [[graph]]
        R1 = foo
[runtime]
    [[foo]]

```

Instructions to reproduce:

```console
$ cylc vip --pause
$ cylc trigger <workflow>//1/foo --flow=2
$ cylc play <workflow>
$ sleep 5
$ sqlite3 ~/cylc-run/one/run1/.service/db 'select * from task_outputs'
1|foo|[1]|["submitted", "started", "succeeded"]
```

The new flow is logged:

```
INFO - New flow: 2 (no description) 
```

However, the task does not run in it:

```
INFO - [1/foo:waiting] => preparing
```

As evidenced by the DB showing the resulting task outputs as existing in flow `[1]`.

### PR DESCRIPTION
**Pull request by hjoliver**
Supersede #6232 
Close #6174 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/data_store_mgr.py [hunk 1]
```diff
@@ -2367,6 +2367,25 @@ def delta_task_queued(self, itask: TaskProxy) -> None:
         self.state_update_families.add(tproxy.first_parent)
         self.updates_pending = True
 
+    def delta_task_flow_nums(self, itask: TaskProxy) -> None:
```
Comment (hjoliver) [inline cylc/flow/data_store_mgr.py:2370]: There was no way to update flow_nums in the data_store. 

We got away with this by (erroneously) re-adding flow-merged n=0 tasks to the task pool, which created an entirely new datastore task proxy.

File: cylc/flow/task_pool.py [hunk 2]
```diff
@@ -1411,10 +1412,12 @@ def spawn_on_output(self, itask, output, forced=False):
                         tasks.append(c_task)
                 else:
                     tasks = [c_task]
+
                 for t in tasks:
                     t.satisfy_me([itask.tokens.duplicate(task_sel=output)])
                     self.data_store_mgr.delta_task_prerequisite(t)
-                    self.add_to_pool(t)
+                    if not in_pool:
+                        self.add_to_pool(t)
```
Comment (hjoliver) [inline cylc/flow/task_pool.py:1420]: This is the add-to-pool bug.

File: cylc/flow/task_pool.py [hunk 3]
```diff
@@ -2169,6 +2172,7 @@ def force_trigger_tasks(
             if itask.state(TASK_STATUS_PREPARING, *TASK_STATUSES_ACTIVE):
                 LOG.warning(f"[{itask}] ignoring trigger - already active")
                 continue
+            self.merge_flows(itask, flow_nums)
```
Comment (hjoliver) [inline cylc/flow/task_pool.py:2175]: This is all that was needed to trigger an existing `n=0` task (e.g. queued) in a new flow. The rest of the PR is just making sure flow_nums change end up in the data store.

**Comment by hjoliver** [comment]
(One SOD test broken, expecting an easy fix though...)

File: tests/integration/test_task_pool.py [hunk 4]
```diff
@@ -2086,3 +2086,27 @@ async def test_set_future_flow(flow, scheduler, start, log_filter):
...
+        assert task.flow_nums == {1}
+
+        # trigger this task even though is already queued in flow 1
+        log.clear()
+        one.pool.force_trigger_tasks([task.identity], '2')
+
+        # the merged flow should continue
+        one.resume_workflow()
+        await complete(one, timeout=2)
+        assert db_select(one, False, 'task_outputs', 'flow_nums') == [('[1, 2]',), ('[1]',)]
```
Comment (oliver-sanders) [inline tests/integration/test_task_pool.py:2112]: Can remove the logging stuff:

```suggestion
async def test_trigger_queue(one, run, db_select, complete,):
    """It should handle triggering tasks in the queued state.

    Triggering a queued task with a new flow number should result in the
    task running with merged flow numbers.

    See https://github.com/cylc/cylc-flow/pull/6241
    """
    async with run(one):
        # the workflow should start up with one task in the original flow
        task = one.pool.get_tasks()[0]
        assert task.state(TASK_STATUS_WAITING, is_queued=True)
        assert task.flow_nums == {1}

        # trigger this task even though is already queued in flow 1
        one.pool.force_trigger_tasks([task.identity], '2')

        # the merged flow should continue
        one.resume_workflow()
        await complete(one, timeout=2)
        assert db_select(one, False, 'task_outputs', 'flow_nums') == [('[1, 2]',), ('[1]',)]
```

**Review by oliver-sanders — APPROVED** [review]
:+1:

**Comment by wxtim** [comment]
Have replicated bug with a runahead limited task - but I still not OK on your branch.

```
[scheduling]
    initial cycle point = 1000
    runahead limit = P2
    [[graph]]
        P1Y = foo
[runtime]
    [[foo]]
        script = sleep 40
```

```
cylc vip
sleep 5
cylc trigger workflow//10030101T0000Z/foo
sleep 40
check db

**Review by wxtim — APPROVED** [review]

**Pull request merged**